modules/table/src/testFixtures/java/org/apache/ignite/internal/table/TxAbstractTest.java

```diff
@@ -2151,12 +2158,178 @@ public void testYoungerTransactionThrowsExceptionIfKeyLockedByOlderTransactionWi
+        testTransactionMultiThreadedFinish(1, false);
+    }
+
+    @RepeatedTest(10)
+    public void testTransactionMultiThreadedCommitEmpty() {
+        testTransactionMultiThreadedFinish(1, true);
+    }
+
+    @RepeatedTest(10)
+    public void testTransactionMultiThreadedRollback() {
+        testTransactionMultiThreadedFinish(0, false);
+    }
+
+    @RepeatedTest(10)
+    public void testTransactionMultiThreadedRollbackEmpty() {
+        testTransactionMultiThreadedFinish(0, true);
+    }
+
+    @RepeatedTest(10)
+    public void testTransactionMultiThreadedMixed() {
+        testTransactionMultiThreadedFinish(-1, false);
+    }
+
+    @RepeatedTest(10)
+    public void testTransactionMultiThreadedMixedEmpty() {
+        testTransactionMultiThreadedFinish(-1, true);
+    }
+
+    /**
+     * Test trying to finish a tx in multiple threads simultaneously, and enlist new operations right after the first finish.
+     *
+     * @param finishMode 1 is commit, 0 is rollback, otherwise random outcome.
+     * @param checkEmptyTx Whether the tx should be empty on finishing (no enlisted operations).
+     */
+    private void testTransactionMultiThreadedFinish(int finishMode, boolean checkEmptyTx) {
+        var rv = accounts.recordView();
+
+        rv.upsert(null, makeValue(1, 1.));
+
+        Transaction tx = igniteTransactions.begin();
+
+        var txId = ((ReadWriteTransactionImpl) tx).id();
+
+        log.info("Started transaction {}", txId);
+
+        if (!checkEmptyTx) {
+            rv.upsert(tx, makeValue(1, 100.));
+            rv.upsert(tx, makeValue(2, 200.));
+        }
+
+        int threadNum = Runtime.getRuntime().availableProcessors() * 5;
+
+        CyclicBarrier b = new CyclicBarrier(threadNum);
+        CountDownLatch finishLatch = new CountDownLatch(1);
+
+        List<Exception> enlistExceptions = synchronizedList(new ArrayList<>());
+
+        var futEnlists = runMultiThreadedAsync(() -> {
+            finishLatch.await();
+            var rnd = ThreadLocalRandom.current();
+
+            try {
+                if (rnd.nextBoolean()) {
+                    rv.upsert(tx, makeValue(2, 200.));
+                } else {
+                    rv.get(tx, makeKey(1));
+                }
+            } catch (Exception e) {
+                enlistExceptions.add(e);
+            }
+
+            return null;
+        }, threadNum, "txCommitTestThread");
+
+        var futFinishes = runMultiThreadedAsync(() -> {
+            b.await();
+
+            finishTx(tx, finishMode);
+
+            finishLatch.countDown();
+
+            return null;
+        }, threadNum, "txCommitTestThread");
+
+        assertThat(futFinishes, willSucceedFast());
+        assertThat(futEnlists, willSucceedFast());
+
+        assertEquals(threadNum, enlistExceptions.size());
+
+        for (var e : enlistExceptions) {
+            try {
+                assertInstanceOf(TransactionException.class, e);
```

Comment:
I believe that we either already have an utility method to fully verify exception with class, code and message or may create it if we don't.

Reply:
assertThrowsWithCause? I think it's not what is needed, it checks a Runnable and doesnt have multiple acceptable messages

Reply:
Why it's possible to have two different messages here?